src/tooltip.tsx

```diff
@@ -112,386 +161,442 @@ const Tooltip: React.FC<Props> = ({
+            case 'right':
+            default:
+                arrowProps = {
+                    top: Math.max(
+                        arrowOffsetFromViewport,
+                        Math.min(
+                            windowSize.height - arrowOffsetFromViewport - ARROW_SIZE,
+                            (top + bottom - ARROW_SIZE) / 2
+                        )
+                    ),
+                    right: '100%',
+                    transform: 'rotate(90deg)',
+                    transformOrigin: 'bottom',
+                };
+
+                break;
+        }
+
+        if (typeof arrowProps.top === 'number') {
+            arrowProps.top -= tooltipProps.top;
+        }
+        if (typeof arrowProps.left === 'number') {
+            arrowProps.left -= tooltipProps.left;
+        }
+
+        if (!isEqual(tooltipProps, tooltipComputedProps)) {
+            setTooltipComputedProps(tooltipProps);
+        }
+        if (!isEqual(arrowProps, arrowComputedProps)) {
+            setArrowComputedProps(arrowProps);
+        }
+    }, [
+        tooltip,
+        targetRect,
+        isTooltipOpen,
+        position,
+        windowSize,
+        tooltipComputedProps,
+        arrowComputedProps,
+        isInverse,
+        isTouchableDevice,
+        tooltipId,
+        resetTooltipInteractions,
+    ]);
+
+    const isTabKeyDownRef = React.useRef(false);
 
     React.useEffect(() => {
-        if (!tooltipRef.current) return undefined;
+        const handleKeyDown = (e: KeyboardEvent) => {
+            switch (e.keyCode) {
+                case ESC:
+                    resetTooltipInteractions();
+                    break;
+                case TAB:
+                    isTabKeyDownRef.current = true;
+                    break;
+                default:
+                // do nothing
+            }
+        };
 
-        tooltipRef.current.getBoundingClientRect();
+        const handleKeyUp = () => (isTabKeyDownRef.current = false);
+
+        const handleOnClick = (e: MouseEvent) => {
+            if (
+                isTouchableDevice &&
+                targetRect &&
+                (e.clientX < targetRect.left ||
+                    e.clientX > targetRect.right ||
+                    e.clientY < targetRect.top ||
+                    e.clientY > targetRect.bottom)
+            ) {
+                resetTooltipInteractions();
+            }
+        };
 
-        if (isVisible) {
-            setTooltipClientRect(tooltipRef.current.getBoundingClientRect());
+        document.addEventListener('keydown', handleKeyDown, false);
+        document.addEventListener('keyup', handleKeyUp, false);
+        document.addEventListener('click', handleOnClick, false);
+        return () => {
+            document.removeEventListener('keydown', handleKeyDown, false);
+            document.removeEventListener('keyup', handleKeyUp, false);
+            document.removeEventListener('click', handleOnClick, false);
+        };
+    });
```

Comment:
I miss the dependency array in this useEffect

Reply:
good catch :)